### PR TITLE
NIFI-1620 Allow empty Content-Type in InvokeHTTP processor

### PR DESCRIPTION
**Pull request by pvillard31**

### CODE REVIEW

**Comment by pvillard31** [comment]
Due to JIRA unavailability, I think the PR has not been linked to issue in JIRA. Just adding a comment to mention the PR in JIRA.

**Comment by taftster** [comment]
I'm not entirely sure if this is a good idea.  Any web service which _disallows_ a standard HTTP header is arguably broken.  Quoting RFC 2616:

>   Any HTTP/1.1 message containing an entity-body SHOULD include a
>   Content-Type header field defining the media type of that body. If
>   and only if the media type is not given by a Content-Type field, the
>   recipient MAY attempt to guess the media type via inspection of its
>   content and/or the name extension(s) of the URI used to identify the
>   resource. If the media type remains unknown, the recipient SHOULD
>   treat it as type "application/octet-stream".

It seems hard to believe given the above that a web service would reject a response with Content-Type.  The current behavior of InvokeHTTP is possibly the most consistent with the HTTP specification.

A custom processor designed to specifically interact with the remote service in question should be considered as an alternative to modifying InvokeHTTP.

**Comment by joewitt** [comment]
i was reviewing this earlier today and frankly had a similar concern to this as Adam.  I didn't reply because I hadn't really figured out what to think.  First, I agree that a service which rejects that header is arguably broken.  Second, as the patch is right now I am curious how it works when the value is empty string because there is a static call to MediaType.... which seems like it would have trouble (still need to verify the logic there though).

However, having said this Pierre can you clarify if the intent is only for the case where there is no entity body or is it also for when there is an entity body in the request?  If the idea is that this is only necessary when there is no entity body we should tighten the code for that case and if it is for either scenario then I think i'm of similar mind to Adam here.

**Comment by pvillard31** [comment]
Definitely agree with you that this API should not act this way.

Anyway, just to clarify, this is only in the case there is no body/content sent with the request.

That's why, in a message on the mailing list, I suggested to add a "body" property. If this property is set (possibly set to empty), the possible (incoming relationship is not mandatory with this processor) incoming flow file is ignored to set content and the request body is constructed with the property. This way we can validate the empty content type if and only if body is set to empty.

Let me know if that makes sense.

**Comment by joewitt** [comment]
Thank you @pvillard31 .  That does make sense and in the case that there is no request body being sent over I am supportive of the notion of not sending the content type header.  This also seems in line with Adam's rfc reference.  You want to tweak this PR then to retain existing content header logic but rather have a 'send body' property or something which when false will not send the body and will not set that header?

**Comment by pvillard31** [comment]
@joewitt yes I can have a look this afternoon. I'll try to propose something before the end of the day.

**Comment by pvillard31** [comment]
Not sure to be able to propose something today, I don't have as much free time as expected. I believe this should be moved to 0.7.0.

**Comment by pvillard31** [comment]
Should be OK now : 
- Added body property
- Reverted default Content-Type and original unit tests
- Authorized no incoming connection with POST/PUT
- Added check : if content-type empty, body must be empty
- Added check : if body empty, content-type must be empty
- Added unit tests
- Tested against the API at the origin of the JIRA

**Comment by taftster** [comment]
This pull request is not strictly dealing with NIFI-1620.  The edge case of suppressing the Content-Type header for an empty message-body makes sense to fix and address. NIFI-1620 is dealing with this edge case.

However, this PR is adding the ability to specify an HTTP request message-body directly as a property of InvokeHTTP.  In my opinion, this is not a desired feature, and I am -1 on this.  A static message-body can always be provided in a dataflow using a tool like ReplaceText; keeping the semantics that InvokeHTTP always reads from flowfile payloads has a good consistency.

Recommend that this PR focus solely on the problem dealing with Content-Type of an empty payload for a POST/PUT operation.  Another JIRA can be opened to discuss the merits of how to provide static content to an HTTP transaction.

**Comment by joewitt** [comment]
@pvillard31 If you stay focused on the goal of NIFI-1620 here and avoid the body manipulation this seems like a good step.  Clearly you found a case worthy of support (null body means unset or empty strong content type).  But, i do share Adam's view that it is best to avoid the content manipulation here.  Do you agree or do you feel like for the use case(s) you envisioned this would be problematic?

Thanks

**Comment by pvillard31** [comment]
I updated the PR with a true/false "send body" property, should be in line with last comments. Let me know if not.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java [hunk 1]
```diff
@@ -215,14 +215,24 @@
...
+            .name("Content-Type")
+            .description("The Content-Type to specify for when content is being transmitted through a PUT or POST. "
+                    + "In the case of an empty value after evaluating an expression language expression, Content-Type defaults to " + DEFAULT_CONTENT_TYPE + "."
+                    + "If and only if body is not sent, Content-Type must be set to empty")
+            .required(false)
+            .expressionLanguageSupported(true)
+            .defaultValue("${" + CoreAttributes.MIME_TYPE.key() + "}")
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(AttributeExpression.ResultType.STRING))
+            .build();
+
```
Comment (taftster) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java]: I'm not sure if we even need this change.  In the case that the PROP_SEND_BODY is false, we should always suppress setting the Content-Type header.  Therefore, I would leave this as-is a required field, with documentation in the PROP_SEND_FIELD property describing the behavior for empty message bodies.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java [hunk 2]
```diff
@@ -215,14 +215,24 @@
...
+    public static final PropertyDescriptor PROP_SEND_BODY = new PropertyDescriptor.Builder()
+            .name("Send body")
+            .description("True if the body content should be sent, false otherwise")
+            .defaultValue("true")
```
Comment (taftster) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java]: We should refer to this as the HTTP spec refers to it.  It's called the "message-body" in the spec, so I think we should use that here as well.  i.e.

```
.name("send-message-body")
.displayName("Send Message Body")
.description("If true, sends the HTTP message body on POST/PUT requests (default).  If false, suppresses the message body and content-type header for these requests.")
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java [hunk 3]
```diff
@@ -444,14 +455,27 @@ public void onPropertyModified(final PropertyDescriptor descriptor, final String
...
+        final String contentType = validationContext.getProperty(PROP_CONTENT_TYPE).getValue();
+        if(contentTypeSet && contentType.isEmpty() && sendBody) {
+            results.add(new ValidationResult.Builder().subject("Content-Type and Send body")
+                    .valid(false).explanation("If Content-Type is set to empty, Send body property must be set to false").build());
+        }
+        if(!sendBody && !contentType.isEmpty()) {
+            results.add(new ValidationResult.Builder().subject("Content-Type and Send body").valid(false).explanation("If body is not sent, Content-Type must be set to empty").build());
+        }
+
         return results;
```
Comment (taftster) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java]: This custom validation rule can be removed if the content-type property is left alone.

**Comment by pvillard31** [comment]
@taftster Thanks Adam. I updated the PR regarding your comments.

**Comment by taftster** [comment]
I can try to look at it tonight. But no problems if you want to step in.

**Comment by JPercivall** [comment]
@taftster I'll let you finish it up tonight if you have time since you've already had eyes on it. If you're not able to, I'll take a look tomorrow.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java [hunk 4]
```diff
@@ -215,14 +215,23 @@
...
-        .build();
+            .name("Content-Type")
+            .description("The Content-Type to specify for when content is being transmitted through a PUT or POST. "
+                    + "In the case of an empty value after evaluating an expression language expression, Content-Type defaults to " + DEFAULT_CONTENT_TYPE)
+            .required(true)
+            .expressionLanguageSupported(true)
+            .defaultValue("${" + CoreAttributes.MIME_TYPE.key() + "}")
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(AttributeExpression.ResultType.STRING))
+            .build();
+
```
Comment (taftster) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java]: The changes on lines 218-226 are just whitespace changes.  Can this be reverted back to the original?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java [hunk 5]
```diff
@@ -761,24 +771,38 @@ private Request configureRequest(final ProcessContext context, final ProcessSess
...
+                @Override
+                public void writeTo(BufferedSink sink) throws IOException {
+                    sink.writeUtf8("");
+                }
+
+                @Override
+                public MediaType contentType() {
+                    return MediaType.parse("");
+                }
+            };
```
Comment (taftster) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/InvokeHTTP.java]: For this 'else' clause, it would probably be better to use a RequstBody static factory.  Something like:
`return RequestBody.create(null, new byte[0]);`

I'm not sure the proper way to specific an empty MediaType, though. I think reading the MediaType javadoc, my hunch is that MediaType.parse("") should return null (since empty string is an invalid type).

**Comment by pvillard31** [comment]
It should be OK, let me know if not.

**Comment by taftster** [comment]
I'm happy with this PR.  Thanks @pvillard31 for all the effort and seeing this through many changes and discussions.

@JPercivall can you still merge this? I'm in between environments right now.

**Comment by JPercivall** [comment]
@taftster thanks for taking another look, yup I can merge this when I get a chance.

**Comment by apiri** [comment]
@taftster @JPercivall I'll aid in the assist and wrap this up as I have been scanning through PRs and can merge in.

**Comment by JPercivall** [comment]
Thanks for the assist @apiri
